### ISSUE TITLE
Git LFS Prune - fatal error : out of memory

### ISSUE THREAD

**Issue by warlokkz** [git-lfs/git-lfs#4195]
Hey Folks! 👋

I've been working on large git-lfs project (currently: project folder size on disk is **~72GB**, `.git` folder size on disk is **~33GB**). It is working fine in day-to-day operations, but every time I run `git lfs prune` I have been encountering a `fatal error: out of memory` that kills the operation.

Just wanted to ask for general guidance in what I should do in this situation. Currently, there is a workaround to simply delete the repo and re-clone to get rid of local files.

I've tried to attach the trace from `GIT_TRACE=1 git-lfs-prune > git-lfs.trace.txt 2&>1` on this issue, but it's too large so here's the snippet of the error:
```
trace git-lfs: RETAIN: 4e48d446ac1a4e47960e0e98d1abbf263e9b46179b2ed1055c4ec26bb3edafe4 via ref b576650b96a20e7c0e2a88bc4ec93c80359bf5e1
trace git-lfs: RETAIN: e040bccbdb310d69a2ce49ac589a738110d0e65f4bb54eff1a55f0b24353986e via ref b576650b96a20e7c0e2a88bc4ec93c80359bf5e1
trace git-lfs: RETAIN: 3e4947515b8b089b58fe55c50ce1d134ca49d73d2acb622ddda04c6b6caa9c43 via ref b576650b96a20e7c0e2a88bc4ec93c80359bf5e1
fatal error: out of memory
fatal error: out of memory

runtime stack:
runtime.throw(0x8f13d4, 0xd)
	C:/Go/src/runtime/panic.go:596 +0x9c
runtime.(*mcache).refill(0xb62480, 0xc800000042, 0xc841d7da20)
	C:/Go/src/runtime/mcache.go:124 +0x12e
runtime.(*mcache).nextFree.func1()
	C:/Go/src/runtime/malloc.go:538 +0x39
runtime.systemstack(0xc042026600)
	C:/Go/src/runtime/asm_amd64.s:327 +0x7e
runtime.mstart()
	C:/Go/src/runtime/proc.go:1132

goroutine 1277 [running]:
runtime.systemstack_switch()
	C:/Go/src/runtime/asm_amd64.s:281 fp=0xc059dd5a68 sp=0xc059dd5a60
runtime.(*mcache).nextFree(0xb62480, 0x42, 0x8ad9e0, 0xc83efb9374, 0x0)
	C:/Go/src/runtime/malloc.go:539 +0xc0 fp=0xc059dd5ac0 sp=0xc059dd5a68
runtime.mallocgc(0x8000, 0x84cca0, 0xc059dd5b01, 0x40d564)
	C:/Go/src/runtime/malloc.go:691 +0x866 fp=0xc059dd5b60 sp=0xc059dd5ac0
runtime.makeslice(0x84cca0, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	C:/Go/src/runtime/slice.go:54 +0x82 fp=0xc059dd5bb0 sp=0xc059dd5b60
io.copyBuffer(0xaff800, 0xc841d7da60, 0xaff700, 0xc841d7da80, 0x0, 0x0, 0x0, 0x8909e0, 0xc841d7da01, 0xc841d7da80)
	C:/Go/src/io/io.go:387 +0x29d fp=0xc059dd5c18 sp=0xc059dd5bb0
io.Copy(0xaff800, 0xc841d7da60, 0xaff700, 0xc841d7da80, 0x2, 0x2, 0xc841d7da40)
	C:/Go/src/io/io.go:360 +0x6f fp=0xc059dd5c78 sp=0xc059dd5c18
io.CopyN(0xaff800, 0xc841d7da60, 0xaff700, 0xc841d7da20, 0x1e0, 0x1e0, 0x40, 0x0)
	C:/Go/src/io/io.go:336 +0x97 fp=0xc059dd5cc8 sp=0xc059dd5c78
github.com/git-lfs/git-lfs/lfs.(*PointerScanner).next(0xc04600e740, 0xc807dac640, 0x28, 0xc048515d20, 0x28, 0xc8418d5380, 0x40, 0x0, 0x0, 0xc044a709b8)
	C:/Users/techn/go/src/github.com/git-lfs/git-lfs/lfs/gitscanner_catfilebatch.go:139 +0x221 fp=0xc059dd5e60 sp=0xc059dd5cc8
github.com/git-lfs/git-lfs/lfs.(*PointerScanner).Scan(0xc04600e740, 0xc807dac640, 0x28, 0x1)
	C:/Users/techn/go/src/github.com/git-lfs/git-lfs/lfs/gitscanner_catfilebatch.go:100 +0x100 fp=0xc059dd5f08 sp=0xc059dd5e60
github.com/git-lfs/git-lfs/lfs.runCatFileBatch.func1(0xc0586f0a30, 0xc04600e740, 0xc0566dacc0, 0xc0566dac00, 0xc05767d5c0, 0xc0566dac60)
	C:/Users/techn/go/src/github.com/git-lfs/git-lfs/lfs/gitscanner_catfilebatch.go:29 +0xbc fp=0xc059dd5fb0 sp=0xc059dd5f08
runtime.goexit()
	C:/Go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc059dd5fb8 sp=0xc059dd5fb0
created by github.com/git-lfs/git-lfs/lfs.runCatFileBatch
	C:/Users/techn/go/src/github.com/git-lfs/git-lfs/lfs/gitscanner_catfilebatch.go:57 +0xc9
```

**System environment**
Device
```
Processor: 	                         Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz, 2592 Mhz, 6 Core(s), 12 Logical Processor(s)
System Type: 	                         x64-based PC
OS Name:	                         Microsoft Windows 10 Enterprise
Installed Physical Memory (RAM):	 32.0 GB
```

**Output of `git lfs env`**
```
git-lfs/2.3.1 (GitHub; windows amd64; go 1.8.3; git 01fef48f)
git version 2.21.0.windows.1

Endpoint=https://artifacts.io/artifactory/lfs-vertical-slice (auth=none)
SSH=git@artifacts.io:artifactory/lfs-vertical-slice
LocalWorkingDir=C:\Users\aanich\Game
LocalGitDir=C:\Users\aanich\Game\.git
LocalGitStorageDir=C:\Users\aanich\Game\.git
LocalMediaDir=C:\Users\aanich\Game\.git\lfs\objects
LocalReferenceDir=
TempDir=C:\Users\aanich\Game\.git\lfs\tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=C:\Users\aanich\Game\.git\lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
GIT_LFS_PATH=C:\Program Files\Git LFS
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

Any help is appreciated, thank you! 🙏

**Comment by bk2204** [comment]
Hey,

Sorry it's taken so long to get back to you.  I'm not sure about what's happening here, since we haven't seen this before.  Could you try with the latest Git LFS client (at the moment, 2.11.0), since that may have some different behavior here?  It's possible we may have fixed the out-of-memory problem without realizing that it was a problem.

**Comment by warlokkz** [comment]
Hey @bk2204, looks like upgrading to at least `2.10` has resolved this issue for others and upgrading `2.11` solved it for me. :) Thanks!

**Comment by bk2204** [comment]
Terrific.  Thanks for testing, and I'm glad you got it working.
